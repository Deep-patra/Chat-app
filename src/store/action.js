export const CHANGE_THEME = "CHANGE_THEME";
export const CHANGE_PROFILE_PICTURE = "CHANGE_PROFILE_PICTURE";
export const CHANGE_FRIENDS_LIST = "CHANGE_FRIENDS_LIST";
export const REMOVE_FRIEND = "REMOVE_FRIEND";
export const CHANGE_ACTIVE_FRIEND = "CHANGE_ACTIVE_FRIEND";
export const CHANGE_ACTIVE_GROUP = "CHANGE_ACTIVE_GROUP";
export const ADD_MESSAGE = "ADD_MESSAGE";
export const CHANGE_SEARCH_RESULT = "CHANGE_SEARCH_RESULT";
export const RECIEVED_MESSAGE = "RECIEVED_MESSAGE";
export const RECIEVED_GROUP_MESSAGE = "RECIEVED_GROUP_MESSAGE";
export const CHANGE_GROUP_LIST = "CHANGE_GROUP_LIST";
export const CHANGE_GROUP_PROFILEPIC = "CHANGE_GROUP_PROFILEPIC";
export const ADD_GROUP = "ADD_GROUP";
export const ADD_GROUP_MEMBER = "ADD_GROUP_MEMBER";
export const ADD_GROUP_MESSAGE = "ADD_GROUP_MESSAGE";
export const REMOVE_GROUP_MEMBER = "REMOVE_GROUP_MESSAGE";
export const REMOVE_GROUP = "REMOVE_GROUP";
export const CHANGE_GROUP_MEMBER_LIST = "CHANGE_GROUP_MEMBER_LIST";
export const SET_USER = "SET_USER";
export const ADD_MESSAGE_LIST = "ADD_MESSAGE_LIST";
export const ADD_GROUP_MESSAGE_LIST = "ADD_GROUP_MESSAGE_LIST";