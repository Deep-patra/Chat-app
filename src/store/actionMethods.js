import {
  CHANGE_THEME,
  CHANGE_PROFILE_PICTURE,
  CHANGE_FRIENDS_LIST,
  REMOVE_FRIEND,
  CHANGE_ACTIVE_FRIEND,
  CHANGE_ACTIVE_GROUP,
  ADD_MESSAGE,
  CHANGE_SEARCH_RESULT,
  RECIEVED_MESSAGE,
  RECIEVED_GROUP_MESSAGE,
  CHANGE_GROUP_LIST,
  CHANGE_GROUP_PROFILEPIC,
  ADD_GROUP,
  ADD_GROUP_MEMBER,
  ADD_GROUP_MESSAGE,
  REMOVE_GROUP_MEMBER,
  REMOVE_GROUP,
  CHANGE_GROUP_MEMBER_LIST,
  SET_USER,
  ADD_MESSAGE_LIST,
  ADD_GROUP_MESSAGE_LIST,
  } from './action';

export const setUser = (user) => ({ type: SET_USER, user });
export const changeTheme = () => ({ type: CHANGE_THEME });
export const changeProfilePicture = (profilePicture) => ({ type: CHANGE_PROFILE_PICTURE, profilePicture });
export const changeFriendsList = (friends) => ({ type: CHANGE_FRIENDS_LIST, friends });
export const removeFriend = (friend) => ({ type: REMOVE_FRIEND, friend });
export const changeActiveFriend = (friendId) => ({ type: CHANGE_ACTIVE_FRIEND, friendId });
export const changeActiveGroup = (groupId) => ({ type: CHANGE_ACTIVE_GROUP, groupId });
export const addMessage = (message) => ({ type: ADD_MESSAGE, message });
export const changeSearchResult = (searchResults) => ({ type: CHANGE_SEARCH_RESULT, searchResults });
export const recievedMessage = (friendId, message) => ({ type: RECIEVED_MESSAGE, friendId, message });
export const recievedGroupMessage = (groupId, message) => ({ type: RECIEVED_GROUP_MESSAGE, groupId, message });
export const changeGroupList = (groups) => ({ type: CHANGE_GROUP_LIST, groups });
export const changeGroupProfilePic = (groupId, profilePic) => ({ type: CHANGE_GROUP_PROFILEPIC, groupId, profilePic });
export const addGroup = (group) => ({ type: ADD_GROUP, group });
export const addGroupMember = (groupId, member) => ({ type: ADD_GROUP_MEMBER, groupId, member });
export const addGroupMessage = (groupId, message) => ({ type: ADD_GROUP_MESSAGE, groupId, message });
export const removeGroupMember = (memberId, groupId) => ({ type: REMOVE_GROUP_MEMBER, groupId, memberId });
export const removeGroup = (groupId) => ({ type: REMOVE_GROUP, groupId });
export const changeGroupMemberList = (groupId, members) => ({ type: CHANGE_GROUP_MEMBER_LIST, groupId, members });
export const addMessageList = (friendId, messages) => ({ type: ADD_MESSAGE_LIST, friendId, messages });
export const addGroupMessageList = (groupId, messages) => ({ type: ADD_GROUP_MESSAGE_LIST, groupId, messages });
