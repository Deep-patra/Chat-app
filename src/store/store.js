import { createStore } from 'redux';
import reducer from './reducer';

const store = createStore(reducer);

store.subscribe(() => console.log("changes are made in the store!"));

export default store;
