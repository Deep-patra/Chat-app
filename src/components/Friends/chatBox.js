import React, { useState, useEffect, useRef } from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { 
  Box,
  TextField,
  IconButton,
  Button,
  Avatar,
  InputAdornment,
  Typography,
  Grow,
} from '@mui/material';
import {
  Send,
  AddPhotoAlternate,
  Download,
  ArrowBack,
 } from '@mui/icons-material';
import { styled } from '@mui/material/styles';
import { addMessage, addMessageList } from '../../store/actionMethods';
import StartWorker from '../Socket/socket';

const MessageBox = styled(Box)(({ theme }) => ({
  border: "solid 2px",
  borderColor: theme.palette.primaryColor.main,
  minWidth: "100px",
  borderRadius: "10px",
  marginLeft: "5px",
  padding: "5px",
  display: "grid",
  gridTemplateRows: "minmax(20px, auto) auto minmax(10px, auto)",
  gridTemplateColumns: "1fr",
  gridRowGap: "4px",
}));

const ImageList = ({ images }) => {
  return (
    <Box minWidth="200px" m="0" disablePadding>
      { images !== undefined ? images.map((image, index) => (
        <Box key={index} minWidth="200px" maxWidth="250px" disablePadding position="relative">
          <img alt="" src={image} width="100%" height="100%" style={{ objectFit: "contain" }} />
          <a
            download="image.jpeg"
            href={image}
            style={{
              position: "absolute",
              bottom: "10px",
              right: "10px",
            }}
          >
            <Download 
              sx={{
                border: "1px solid",
                borderColor: "primaryColor.main",
                backgroundColor: "#fff",
                borderRadius: "50%",
              }}
            />
          </a>
        </Box> 
      )) : null }
    </Box>
  );
}; 


const Message = ({ message, user, friend }) => {
  const { authorId, text, images, time } = message;
  return (
    <Grow in={true}>
      <Box
        display="flex"
        flexDirection="row"
        flexWrap="nowrap"
        marginY="5px"
        maxWidth="400px"
      >
        <Avatar
          alt={""}
          src={user.userId === authorId ? user.profilePicture : friend.profilePicture }
          sx={(theme) => ({
            width: "40px",
            height: "40px",
          })}
        >
          { user.userId === authorId ? user.name.charAt(0) : friend.friendName.charAt(0) }
        </Avatar>
        <MessageBox>
          <Typography sx={(theme) => ({ color : (theme.palette.mode === 'light' ? 'primaryColor.main' : '#fff') })}>{authorId === user.userId ? user.name : friend.friendName }</Typography>
          {images !== undefined && images.length !== 0 ?
            <ImageList images={images} />
            : null }
          {text !== undefined && text !== '' ?
            <Typography>{text}</Typography>
            : null}
          <Box display="flex" flexDirection="row" justifyContent="flex-end">
            <Typography sx={{ color: "primaryColor.main" }} fontSize="10px" component="span" >{time}</Typography>
          </Box>
        </MessageBox>
      </Box>
    </Grow>
  );
};


// convert the input files to blob for preview
const convertImage = (files, changeFile) => {
  // console.log(files);
  if (files.length > 0) {
    const fileList = [];
    for (const file of files) {
      fileList.push(URL.createObjectURL(file));
      // console.log(fileList)
    }

    changeFile([...fileList]);
  }
};

const getBase64String = (file) => new Promise((resolve, reject) => {
  const canvas = document.createElement('canvas');
  const ctx = canvas.getContext('2d');

  const img = new Image();

  img.onload = () => {
    canvas.width = 200;
    canvas.height = 200;

    ctx.drawImage(img, 0, 0, 200, 200);

    const image = canvas.toDataURL('jpeg', 0.5);
    resolve(image);
  };

  img.onerror = (error) => {
    reject(error);
  };

  img.src = URL.createObjectURL(file);
});

const convertImageToBase64 = async (files) => {
  const base64Files = [];
  if (files.length === 1 && files[0].name === "a.txt") return [];
  if (files.length > 0) {
    for (const file of files) {
      base64Files.push(await getBase64String(file));
    }
  }
  return base64Files;
}



const TypingBox = ({ userId, friendId }) => {
  const inputRef = useRef(null);
  const dispatch = useDispatch();
  const [ value, changeValue ] = useState('');
  const [ files, changeFile ] =useState([]);

  const handleInputChange = (e) => {
    changeValue(e.target.value);
  };

  const handleSendButton = async () => {
    if (value !== "" || files.length > 0) {
      console.log(inputRef.current.files);
      const imageFiles = await convertImageToBase64(inputRef.current.files);
      console.log(imageFiles);
      dispatch(addMessage({
        authorId: userId,
        text: value || '',
        images: imageFiles || [],
        time: `${new Date().toLocaleTimeString()}`,
      }));

      // sending the message from the websocket
      StartWorker.handleAddMessage(friendId, value, imageFiles);
      
      // reset the values and the files in the state
      changeValue('');
      changeFile([]);
      // reseting the files property in input element
      const dt = new DataTransfer();
      dt.items.add(new File([], 'a.txt'));
      inputRef.current.files = dt.files;
    }
  };

  const handleFileChange = (e) => {
    convertImage(e.target.files, changeFile);
  };

  useEffect(() => {

    const input = inputRef.current;

    input.addEventListener("change", handleFileChange);
    return () => {
      input.removeEventListener("change", handleFileChange);
    };
  }, [inputRef]);

  return (
    <Box display="flex" flexDirection="row" flexWrap="nowrap">
      <input
        type="file"
        accept="image/*"
        multiple
        ref={inputRef}
        style={{ display: "none" }}
      />
      <TextField
        variant="outlined"
        type="text"
        value={value}
        placeholder="Type Something..."
        onChange={handleInputChange}
        autoFocus
        color="primaryColor"
        sx={{ width: "100%" }}
        InputProps={{
          endAdornment: <InputAdornment position="end">
            <IconButton onClick={() => {inputRef.current.click();}}>
              <AddPhotoAlternate sx={{ color: "primaryColor.main" }} />
            </IconButton>
          </InputAdornment>
        }}
      />
      <Button variant="contained" color="primaryColor" onClick={handleSendButton}>
        <Send sx={{color: "#fff"}} />
      </Button>
    </Box>
  );
};

const getActiveFriend = (state) => state.activeFriendId;

const ChatBox = ({ toggleChatBox }) => {

  const messageBoxRef = useRef(null);
  const messageContainerRef = useRef(null);
  const dispatch = useDispatch();

  const friendId = useSelector(getActiveFriend);
  const { friends, messageList } = useSelector((state) => (
    {
      friends: state.friends,
      messageList: state.messages
    }));

  const friend = friends.find((item) => friendId === item.friendId );
  const { friendName, profilePicture } = friend;
  // console.log(friendId, friends, friend);
  // console.log(messageList);
  const item = messageList.find((item) => item.friendId === friendId);

  let messages = [];
  if (item !== undefined && item.messages !== undefined) messages = item.messages;

  // storing the user info in an object
  const user = useSelector((state) => (
    { name: state.name, profilePicture: state.profilePicture, userId: state.userId }
  ));


  useEffect(() => {

    const fetchMessages = () => {
       // fetch the messages
      fetch(`/${user.userId}/friends/${friendId}/messages`, {
        method: 'GET',
        accept: 'application/json',
      })
      .then(response => {
        if (!response.ok || response.status !== 200)
          throw new Error('response is not ok');
        return response;
      })
      .then(response => response.json())
      .then(res => {
        const { messages } = res;

        if (messages === undefined)
          throw new Error('response is not undefined');

          
        // dispatch the messages to the store
        dispatch(addMessageList(friendId, messages));
      })
    };

    fetchMessages();
   
  }, []);


  useEffect(() => {
    const container = messageContainerRef.current.getBoundingClientRect();
    messageBoxRef.current.scroll({
      top: container.height,
      left: 0,
      behavior: "smooth",
    });
  }, [messages.length]);

  return (
    <Box
      display="grid"
      gridTemplateRows="[line1] 50px [line2] calc(100vh - 205px) [line3] 60px [line4]"
      gridTemplateColumns="1fr"
      width={{ xs: "100%" }}
      sx={{ gridGap: "10px" }}
      height="100%"
      disablePadding
    >
      <Box
        gridRow="line1/line2"
        display="flex"
        flexDirection="row"
        flexWrap="nowrap"
        borderBottom="1px solid"
        borderColor="opposite.main"
        pb="5px"
        pt="2px"
        alignItems="center"
      >
        <IconButton onClick={toggleChatBox}>
          <ArrowBack color="primaryColor" />
        </IconButton>
        <Box display="flex" flexDirection="row">
          <Avatar src={profilePicture} sx={{ width: "30px", height: "30px" }} />
          <Typography sx={{ color: "opposite.main", fontSize: "1.3rem", marginLeft: "20px" }} component="span">{friendName}</Typography>
        </Box>
      </Box>
      <Box
        gridRow="line2/line3"
        ref={messageBoxRef}
        minHeight="200px"

        sx={{
          overflowY: "scroll",
          '&::-webkit-scrollbar': {
            bgcolor: "primary.main",
            width: "5px",
            height: "5px",
          },
          '&::-webkit-scrollbar-thumb': {
            maxWidth: "5px",
            bgcolor: "opposite.main",
            borderRadius: "8px",
            '&:hover': {
              bgcolor: "primaryColor.main"
            },
          }
        }}
      >
        <Box
          display="flex"
          flexDirection="column"
          flexWrap="nowrap"
          ref={messageContainerRef}
        >
          {
            messages.map((message, index) => (
            <Message
              key={index}
              message={message}
              friend={friend}
              user={user}
            />
          ))}
          { messages.length === 0 ? 
            <Box display="flex" justifyContent="center" alignItems="center">
              <Typography variant="poster"> No conversation yet! </Typography>
            </Box>
          : null }
        </Box>
      </Box>
      <Box
        gridRow="line3/line4"
        display="flex"
        flexDirection="column"
        justifyContent="flex-end"
        disablePadding
      >
        <TypingBox
          userId={user.userId}
          friendId={friendId}
        />
      </Box>
    </Box>
  );
};

export default ChatBox;
