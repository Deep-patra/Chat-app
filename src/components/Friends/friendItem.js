import React, { useState, Fragment } from 'react';
import { 
  ListItem,
  ListItemAvatar,
  Avatar,
  ButtonGroup,
  ListItemText,
  IconButton,
  Typography,
  Popover,
  Tooltip,
  Divider,
  Button,
 } from "@mui/material";
import { 
  MoreVert,
  Chat,
  Delete,
 } from "@mui/icons-material";
 import { useDispatch } from 'react-redux';
 import { changeActiveFriend } from '../../store/actionMethods';


const FriendItem = ({ friend, handleRemove, handleClick }) => {
  const [anchorEl, setAnchorEl] = useState(null);
  const { friendName, friendId, profilePicture } = friend;
  const dispatch = useDispatch();

  const handleMenuButton = (e) => {
    setAnchorEl(e.currentTarget)
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const handleRemoveButton = () => {
    handleRemove(friend);
    setAnchorEl(null);
  };

  const handleChatClick = (e) => {
    dispatch(changeActiveFriend(friendId));
    handleClick();
  };

  const isOpen = Boolean(anchorEl);
  return (
    <Fragment key={friendName} >
      <ListItem disablePadding sx={{ pb: 1, pt: 1}} >
        <ListItemAvatar>
          <Avatar alt={friendName} src={profilePicture}>{profilePicture === '' ? friendName.charAt(0) : null}</Avatar>
        </ListItemAvatar>
        <ListItemText
          primary={
            <Fragment>
              <Typography
                component="span"
              >
                {friendName}
              </Typography>
            </Fragment>
          }
        />
      <ButtonGroup variant="outlined">
        <Tooltip title="start conversation">
          <IconButton onClick={handleChatClick}>
            <Chat sx={{ color: "primaryColor.main" }} />
          </IconButton>
        </Tooltip>
        <IconButton onClick={handleMenuButton}>
          <MoreVert sx={{ color: "opposite.main" }} />
        </IconButton>
        <Popover
          open={isOpen}
          anchorEl={anchorEl}
          onClose={handleClose}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
        >
          <Button
            variant="text"
            onClick={handleRemoveButton}
            color="opposite"
            sx={{ fontSize: '0.7rem' }}
          >
            <Delete sx={{ color: "error.main" }} fontSize="small" />
            Remove
          </Button>
        </Popover>
      </ButtonGroup>
      </ListItem>
      <Divider/>
    </Fragment>
  );
};

export default FriendItem;