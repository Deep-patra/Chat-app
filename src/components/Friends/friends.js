import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Box,
  List,
  IconButton,
  Typography,
  Tooltip,
  ButtonGroup,
  CircularProgress,
} from '@mui/material';
import { styled } from '@mui/material/styles';
import { Error, AddBox, CancelScheduleSend, People, } from '@mui/icons-material';
import { changeFriendsList , removeFriend } from '../../store/actionMethods';
import FriendItem from './friendItem';
import ChatBox from './chatBox.js';



// fetching friends list from the server
const fetchFriends = async (userId) => {
  const response = await fetch(`/${userId}/friends`, {
    method: 'GET',
    header: { 'Content-Type': 'application/json' },
    accept: 'application/json',
  })
  .catch( error => console.log(error));

  if (response === undefined) throw new Error ('response is invalid!');
  if (!response.ok) throw new Error('response is not ok');
  if (response.status !== 200) throw new Error('response status is not 200');

  const res = await response.json();
  if (res.friends === undefined) return [];
  return res.friends;
};

// funtion to get the friends list from the redux store
const getUserId = (state) => state.userId;


const FriendsBox = ({ toggleChatBox, handleClick }) => {

  let userId = useSelector(getUserId);
  const dispatch = useDispatch();
  let friends = useSelector(state => state.friends);
  const [ loading, changeLoadingStatus ] = useState(false);
  const [activeButtons, buttonClicked] = useState({
    friendsOpen: true,
    receivedRequest: false,
    sentRequests: false,
  });

  useEffect(() => {
    if ( Array.isArray(friends) && friends.length === 0 ) {

      // play the loading the animation
      changeLoadingStatus(true);
  
      fetchFriends(userId)
      .then(result => {
        const res = result.friends;
        dispatch(changeFriendsList(res));
        // stop the animation
        changeLoadingStatus(false);
      })
      .catch(err => {
        changeLoadingStatus(false);
      });
    }
  
  }, []);
  // console.log(friends);
  // for testing the component
  // comment this line in productions mode
  // friends = [{ name: "hello", profilePic: "" }];

  // handle the remove button 
  const handleRemoveButton = (friend) => {
    dispatch(removeFriend(friend));
  };


  return (
    <>
      <Box display="flex" flexDirection="row" flexWrap="nowrap" justifyContent="space-between">
          <Typography component="span" variant="poster" sx={{fontSize: '2rem'}}>Friends</Typography>
          <ButtonGroup variant="outlined" size="small" color="primaryColor" aria-label="outline button group">
            <Tooltip title="friends">
              <IconButton>
                <People sx={{color: (activeButtons.friendsOpen ? "primaryColor.main" : "opposite.main" )}} />
              </IconButton>
            </Tooltip>
            <Tooltip title="Friend Requests">
              <IconButton>
                <AddBox sx={{color: (activeButtons.receivedRequest ? "primaryColor.main" : "opposite.main")}} />
              </IconButton>
            </Tooltip>
            <Tooltip title="Sent Requests">
              <IconButton>
                <CancelScheduleSend sx={{ color: (activeButtons.sentRequests ? "primaryColor.main" : "opposite.main" ) }}/>
              </IconButton>
            </Tooltip>
          </ButtonGroup>
        </Box>
        {
          loading ?
            <Box display="flex" flexDirection="column" flexWrap="nowrap" alignItems="center" marginTop="100px">
              <CircularProgress sx={{ color: 'opposite.main' }} size="1.5rem" />
            </Box>
          : <List>
              { friends.length !== 0 && friends !== undefined ? [...friends].map((friend, index) => (
                <FriendItem friend={friend} handleRemove={handleRemoveButton} handleClick={handleClick} key={index} />
              )) : null }
            </List>
        }
        { !loading && friends.length === 0 ?
          <Box
            display="flex"
            flexDirection="column"
            flexWrap="nowrap"
            justifyContent="center"
            alignItems="center"
          >
            <Error sx={{ color: 'error.main' }} />
            <Typography variant="poster" component="span">No Friends yet!</Typography>
          </Box>
        : null
        }
    </>
  );
};


const Friends = () => {
  const [ isOpen, toggleChatBox ] = useState(false);

  const openChatBox = () => {
    toggleChatBox(true);
  };

  const closeChatBox = () => {
    toggleChatBox(false);
  };

  return (
    <Box
      display="flex"
      flexDirection="column"
      flexWrap="nowrap"
      gridColumn={{
        sm: "2/ span 10",
        md: "2/ span 8",
      }}
      height="100%"
      maxWidth="800px"
      p={{ xs: "10px" }}
      pt={{ xs: "0px" }}
      pl={{ sm: "30px", md: "50px" }}
    >
      { !isOpen ? 
        <FriendsBox toggleChatBox={toggleChatBox} handleClick={openChatBox} /> :
        <ChatBox toggleChatBox={closeChatBox} />
      }
    </Box>
  );
};

export default Friends;