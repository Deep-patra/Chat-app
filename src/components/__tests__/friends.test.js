import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { CHANGE_FRIEND_LIST } from '../../store/action';
import Friends from '../Friends/friends';
import { render, screen, cleanup } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

const server = setupServer(
  rest.get('/:userId/friends', (req, res, ctx) => {
    return ctx.json({ friends: [
      {
        name: 'XYZ',
        profilePicture: 'yauhd'
      },
      {
        name: 'ABC',
        profilePicture: 'shfdfkf',
      }
    ] });
  }),
);

const initialState = {
  userId: '124566',
  friends: [],
};

const reducer = (state, action) => {
  const { type, friends } = action;
  if (type === CHANGE_FRIEND_LIST) {
    return { ...initialState, friends };
  }
  return { ...initialState };
}

const store = createStore(reducer);

beforeAll(() => server.listen());
afterEach(() => { server.resetHandlers(); cleanup(); });
afterAll(() => server.close());

test('Testing the friends component', () => {
  const { container } = render(
    <Provider store={store}>
      <Friends/>
    </Provider>
  );

  expect(container).toMatchSnapshot();
});

