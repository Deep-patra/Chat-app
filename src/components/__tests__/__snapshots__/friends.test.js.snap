// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the friends component 1`] = `
<div>
  <div
    class="MuiBox-root css-zwhokb"
  >
    <div
      class="MuiBox-root css-1gts8n"
    >
      <span
        class="MuiTypography-root MuiTypography-poster css-92zgu-MuiTypography-root"
      >
        Friends
      </span>
      <div
        aria-label="outline button group"
        class="MuiButtonGroup-root MuiButtonGroup-outlined css-iajp3t-MuiButtonGroup-root"
        role="group"
      >
        <button
          aria-label="friends"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-cly8x9-MuiSvgIcon-root"
            data-testid="PeopleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5c-1.66 0-3 1.34-3 3s1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5C6.34 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Friend Requests"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1hfhy7y-MuiSvgIcon-root"
            data-testid="AddBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Sent Requests"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-mui-internal-clone-element="true"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-cly8x9-MuiSvgIcon-root"
            data-testid="CancelScheduleSendIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.5 9c-.42 0-.83.04-1.24.11L1.01 3 1 10l9 2-9 2 .01 7 8.07-3.46C9.59 21.19 12.71 24 16.5 24c4.14 0 7.5-3.36 7.5-7.5S20.64 9 16.5 9zm0 13c-3.03 0-5.5-2.47-5.5-5.5s2.47-5.5 5.5-5.5 5.5 2.47 5.5 5.5-2.47 5.5-5.5 5.5z"
            />
            <path
              d="m18.27 14.03-1.77 1.76-1.77-1.76-.7.7 1.76 1.77-1.76 1.77.7.7 1.77-1.76 1.77 1.76.7-.7-1.76-1.77 1.76-1.77z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-1p6nnf3"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-9c8soh-MuiCircularProgress-root"
        role="progressbar"
        style="width: 1rem; height: 1rem;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
      <span
        class="MuiTypography-root MuiTypography-poster css-37jdci-MuiTypography-root"
      >
        Fetching your friend List
      </span>
    </div>
  </div>
</div>
`;
